### PR TITLE
Adding User support for Detector and DetectorJob

### PR DESCRIPTION
**Pull request by saratvemulapalli**
*Issue #195 *

*Description of changes:*
Adding User support to store User information for each detector created and store user context in .opendistro-anomaly-detectors index.
Adding User support for AD Job and store user context in .opendistro-anomaly-detectors-jobs.

User role will be used to filter when AD Rest API's are invoked. 
Background jobs will soon have injected role from the User context. 

User information is always sent to elasticsearch when Opendistro security plugin is enabled.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by ohltyler — APPROVED** [review]

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java [hunk 1]
```diff
@@ -289,6 +308,9 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         if (categoryFields != null) {
             xContentBuilder.field(CATEGORY_FIELD, categoryFields.toArray());
         }
+        if (user != null) {
```
Comment (ylwu-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:311]: Add `user` field in detector index mapping https://github.com/opendistro-for-elasticsearch/anomaly-detection/blob/master/src/main/resources/mappings/anomaly-detectors.json ?
Same to detector job.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java [hunk 2]
```diff
@@ -131,7 +135,8 @@ public AnomalyDetector(
         Map<String, Object> uiMetadata,
         Integer schemaVersion,
         Instant lastUpdateTime,
-        List<String> categoryFields
+        List<String> categoryFields,
+        User user
```
Comment (ylwu-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:139]: Here we just add `User` in each detector document. Security plugin will do authorization check for each request based on `backend_roles` or `roles` ? Can you share the related code?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java [hunk 3]
```diff
@@ -260,6 +273,12 @@ public void writeTo(StreamOutput output) throws IOException {
         output.writeInt(schemaVersion);
         output.writeInstant(lastUpdateTime);
         output.writeStringCollection(categoryFields);
+        if (user != null) {
+            output.writeBoolean(true); // user exists
+            user.writeTo(output);
+        } else {
+            output.writeBoolean(false); // user does not exist
```
Comment (ylwu-amzn) [inline src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:280]: For old detector, they have no user field, so they will be open to all users which has AD permission ?
For new detector, we will create default user with the creator's user role?

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java (hunk 1 above)
Comment (saratvemulapalli) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:311]: Great point sure. Will add.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java (hunk 2 above)
Comment (saratvemulapalli) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:139]: Good question, security plugin will only provide the authorization headers.
My next PR will address this. Basically we have to query and filter by `roles` or `backend_roles`.
Here is a code pointer in alerting which does that: https://github.com/opendistro-for-elasticsearch/alerting/commit/44abca11b49a658fc22b25ab72c4ab62bd9057f3#diff-39bec5249e2b91e055cdb30bb3edcd28ee868ab3676e98652547bcce28eed895R86

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java (hunk 3 above)
Comment (saratvemulapalli) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:280]: Exactly, for existing detectors where User doesn't exist, the filter will not work and users who have all access to rest api's will be able to see those detectors. 
This is to maintain backward compatibility.

File: src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java (hunk 2 above)
Comment (ylwu-amzn) [reply src/main/java/com/amazon/opendistroforelasticsearch/ad/model/AnomalyDetector.java:139]: Got it, that makes sense.

**Review by ylwu-amzn — APPROVED** [review]
LGTM. Thanks for the change!

**Review by skkosuri-amzn — COMMENTED** [review]
Looks good. 👍

**Pull request merged**